### PR TITLE
[quant] fix dropout in FX graph mode quantization

### PR DESCRIPTION
**Pull request by terrychenism**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71043

Summary:
fix issue #68250
dropout break fx graph model quantization

Test Plan:
python test/test_quantization.py TestStaticQuantizedModule

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33490176](https://our.internmc.facebook.com/intern/diff/D33490176)

### CODE REVIEW

**Comment by jerryzh168** [comment]
I think we probably don't need qat module, the support is similar to quantized batchnorm. we can add a test similar to batchnorm: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_module.py#L611 as well in the same TestCase

**Comment by jerryzh168** [comment]
Also would be good to include the test in the issue as well, we can add a test function in https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L5536

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

File: torch/nn/quantized/modules/dropout.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    Args:
+        scale: quantization scale of the output tensor
+        zero_point: quantization zero point of the output tensor
+        p: probability of an element to be zeroed
+        inplace: can optionally do the operation in-place. Default: ``False``
+    """
+    def __init__(self, p=0.5, inplace=False):
+        super(Dropout, self).__init__(p, inplace)
+        self.inplace = inplace
```
Comment (vadimkantorov) [inline torch/nn/quantized/modules/dropout.py:15]: are these lines actually needed? the base class does set these fields, right? also, `super().__init__(p, inplace)` is a bit more modern / less verbose...

File: test/quantization/fx/test_quantize_fx.py [hunk 2]
```diff
@@ -5617,6 +5617,15 @@ def forward(self, x):
                 out = model_quantized(input.to(device_after))
                 self.assertEqual(out.device.type, device_after)
 
+    def _test_model_dropout(self):
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:5620]: nit: this looks like a normal test instead of a helper function, probably don't need to start with `_`

**Comment by jerryzh168** [comment]
Looks good to me, but I'll let @vkuzo to take a final look

File: torch/nn/quantized/modules/dropout.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import torch
+import torch.nn.quantized.functional
+
+class Dropout(torch.nn.Dropout):
```
Comment (vkuzo) [inline torch/nn/quantized/modules/dropout.py:4]: Thanks for fixing this.  It would be valuable to add some context to why the logic of this class is not actually doing dropout, i.e. there is currently no need or implementation for actual quantized dropout, and this is a placeholder to enable models where fp32 tensors had dropout to work with quantized tensors.

Comment (vkuzo) [reply torch/nn/quantized/modules/dropout.py:4]: I'm also still not sure why not we just use `nn.Identity` instead of creating this module, but I don't feel super strongly

Comment (terrychenism) [reply torch/nn/quantized/modules/dropout.py:4]: I don't have any preference.  
@jerryzh168 any thoughts?

File: test/quantization/fx/test_quantize_fx.py (hunk 2 above)
Comment (jerryzh168) [reply test/quantization/fx/test_quantize_fx.py:5620]: I think we can add another similar test in eager mode test as well: https://github.com/pytorch/pytorch/blob/master/test/quantization/eager/test_quantize_eager_qat.py#L246

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33490176).

File: torch/nn/quantized/modules/dropout.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    r"""This is the quantized equivalent of :class:`~torch.nn.Dropout`.
```
Comment (vkuzo) [inline torch/nn/quantized/modules/dropout.py:5]: would be good to explicitly say "this class does not actually do dropout"

**Review by vkuzo — APPROVED** [review]
thanks for working on this!

**Comment by datumbox** [comment]
I think this PR broke TorchVision's pre-trained quantized models. See here for all details: https://github.com/pytorch/vision/issues/5198
